Marketplace: "search on hit enter" search component

Rolilink opened the pull request:
#### Changes proposed in this Pull Request
* Adds search modes to the `Search` component.
* Adds the `when-typing` and `on-enter` modes.
* Changes marketplace search bar mode to `on-enter`.

#### Testing instructions

* Go to `/plugins/:siteId`
* Write any search input into the search bar, it should not search when typing.
* Press enter and it should correctly search.
* Click the close button (the one with the X in the search bar), it should clean the search box and go back to the initial state.
* Test that the search bars at `/start/domains` and `/new/language-modal` still work as intended (search as you type).

Related to #62229
Rebase when #62532 is merged

Rolilink committed:
adds sticky search bar to marketplace

Rolilink committed:
refs fixes and useSticky fix

Rolilink committed:
cleanup and docs

Rolilink committed:
adds "on-enter" search mode to the Search component

Rolilink committed:
adds "on-enter" search mode to marketplace search

Rolilink committed:
add documentation to the Search component

Rolilink committed:
add unit tests to "on-enter" search mode

Rolilink committed:
small cleanup

lsl commented on packages/search/src/search.tsx:
```diff
@@ -75,6 +75,7 @@ type Props = {
 	recordEvent?: ( eventName: string ) => void;
 	searching?: boolean;
 	value?: string;
+	searchMode?: string;
```
```suggestion
	searchMode?: 'when-typing' | 'on-enter';
```

lsl commented on packages/search/src/search.tsx:
```diff
@@ -291,6 +311,11 @@ const InnerSearch = (
+		if ( event.key === 'Enter' && searchMode === 'on-enter' ) {
+			bypassOnSearchDebounce( keyword );
+			onSearchChange?.( keyword );
+		}
```
@vinimotaa are we intentionally only supporting `Enter` key presses as a way to submit a search?

vinimotaa replied:
hey @lsl this is what we discussed - the pending decision is to add or not to add a "trigger" button to activate the search. I'd like to hear from you all your opinions. I know for mobile we don't need, this would only be for desktop

paulopmt1 replied:
Hey @vinimotaa. I liked we removed the "search while typing" feature because that normally made me anxious to type fast :)

Maybe because I was used to it, I'm now missing the auto search, at least when I complete typing my search term. If this is intentional, I would suggest adding a trigger button to make it explicit to the user they need to do something to start the search.

vinimotaa replied:
Thanks for your feedback @paulopmt1!

Good point - on this case, what I have in mind for the desktop version is to have something like this;

<img width="1394" alt="image" src="https://user-images.githubusercontent.com/97929097/162214496-d7570882-062e-4b50-b7ac-f5ae5f84cba8.png">

we could incorporate a CTA within the search bar. This would be something I'd like to get some feedback from @SaxonF as well!

Rolilink replied:
@vinimotaa I also would think maybe replacing the search icon with the close button

lsl replied:
Replacing the icon once there is text would be pretty nice but I think we're already borderline implementing two search components in one, might be better to drop it to the left of the Search button, it should look ok.

lsl commented:
@Rolilink Was trying to review understand how this works and came up with a modified version in https://github.com/Automattic/wp-calypso/pull/62581 merge into this one if you're prefer that approach instead.

lsl reviewed (approved):
Please consider [this patch](https://github.com/Automattic/wp-calypso/pull/62581/files), otherwise lgtm.

paulopmt1 reviewed (approved):
Good job @Rolilink!

- I liked your test, it's clear it will cover new component behavior! 
- I also like the way we reuse the search component. Normally on other companies I just saw imports of a given component directly on its files, here I see we're really prepared to reuse things, that's great!

LGTM, I also suggest merging @lsl [code](https://github.com/Automattic/wp-calypso/pull/62581/files) and did one comment for @vinimotaa [above](https://github.com/Automattic/wp-calypso/pull/62573#discussion_r845123527).

lsl committed:
Avoid debounce instead (#62581)

lsl commented on client/my-sites/plugins/search-box-header/index.jsx:
```diff
@@ -20,8 +20,9 @@ const SearchBox = ( { isMobile, doSearch, searchTerm, searchBoxRef } ) => {
+				searchMode="on-enter"
 				placeholder={ translate( 'Try searching "ecommerce"' ) }
-				delaySearch={ true }
+				delaySearch={ false }
```
```suggestion
```
I think this might be an option now. I think I changed this thinking I'd use doSearch internally but it made more sense to use onSearch directly for the "on-enter" mode.

jsnajdr commented on packages/search/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
 
 - Add Search component based on `calypso/components/search`
+
+## 1.0.1
```
New changelog entries should be at the beginning, so that the latest version is first.

jsnajdr commented on packages/search/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+## 1.0.1
+
+- Add search modes
+- Add "when-typing" search mode
+- Add "on-enter" searchmode.
```
Can be consolidated to one 'Add "on-enter" and "when-typing" search modes.' entry.

lsl replied:
```suggestion
- Add "when-typing" and "on-enter" search modes, defaulting to "when-typing".
```

jsnajdr commented on packages/search/package.json:
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "@automattic/search",
-	"version": "1.0.0",
+	"version": "1.0.1",
```
As there are new features, the version bump should be to 1.1.0.

jsnajdr commented on packages/search/src/search.stories.jsx:
```diff
@@ -1,3 +1,5 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { action } from '@storybook/addon-actions';
```
ESLint probably complains here because the dependency is not declared in package's `devDependencies`? Adding it there should remove the need for disabling warnings.

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -228,6 +234,10 @@ const InnerSearch = (
+		if ( 'on-enter' === searchMode ) {
+			onSearch?.( '' );
+			onSearchChange?.( '' );
+		}
```
Why trigger a search when the search box is being closed?

lsl replied:
Probably needs to be re-checked but I think this is so when you click the "X" clear/reset button the results fallback to default empty search.

lsl replied:
Confirming above

lsl committed:
Update packages/search/CHANGELOG.md

Rolilink committed:
Updating changelog
Upgrading to minor version
Adding storybook-actions mising development dependency

Rolilink committed:
dedupe @storybook/addon-actions

lsl commented on packages/search/src/search.stories.jsx:
```diff
@@ -1,3 +1,5 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
```
```suggestion
```

lsl replied:
```suggestion
```

lsl committed:
Update packages/search/src/search.stories.jsx

jsnajdr commented on packages/search/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 1.0.1
+
+- Add search modes
+- Add "when-typing" and "on-enter" search mode
```
Still can be shortened to just one line. Currently both entries say the same thing: search modes were added.
